### ISSUE TITLE
Label time range

### ISSUE THREAD

**Issue by zmlapo** [maxjcohen/transformer#5]
If the input is 8 points between t-7 and t, should the corresponding labels be for times t+1 to t+8?

What does the output time window look like?

**Comment by maxjcohen** [comment]
Hi, sorry i don't quite understand your question. We are dealing here with a coherent many to many problem, so for each time step t, their are one input and one output.

Where is this 8 points coming from ?

**Comment by zmlapo** [comment]
Hi, 8 is just an arbitrary number of points. 

I'm curious about using your transformer for prediction tasks, so that given a set of input data, train it to make predictions about future states.

I suppose I could feed N data points starting at time t as input and then feed N labels starting at time t+1 to train for one step ahead prediction.

**Comment by maxjcohen** [comment]
Ok, this is a very relevant question that we've been asking ourselves. The Transformer computes output in parallel, ie we compute <img src="http://www.sciweavers.org/tex2img.php?eq=%28%5Cwidehat%20Y_1%2C%20%5Ccdots%2C%20%5Cwidehat%20Y_T%29%20%3D%20F%28X_1%2C%20%5Ccdots%2C%20X_T%29&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=0" align="center" border="0" alt="(\widehat Y_1, \cdots, \widehat Y_T) = F(X_1, \cdots, X_T)" width="228" height="22" />, instead of computing at each step <img src="http://www.sciweavers.org/tex2img.php?eq=%5Cwidehat%20Y_t%20%3D%20F%28X_t%2C%20h_%7Bt-1%7D%29&bc=White&fc=Black&im=jpg&fs=12&ff=arev&edit=0" align="center" border="0" alt="\widehat Y_t = F(X_t, h_{t-1})" width="128" height="22" />like an LSTM for example.

For this reason, using Transformer as is to predict one step at a time is not very computationally effective. That being said, [WaveNet](https://www.semanticscholar.org/paper/WaveNet%3A-A-Generative-Model-for-Raw-Audio-Oord-Dieleman/df0402517a7338ae28bc54acaac400de6b456a46) famously had the same issue, and was able to bridge the computation gap by using [Knowledge Distillation](https://www.semanticscholar.org/paper/Parallel-WaveNet%3A-Fast-High-Fidelity-Speech-Oord-Li/f6cbf83e1ce3b099d656d2346b261d5ef7f2b62e).

**Comment by zmlapo** [comment]
Thanks Max, I am running some experiments now.

**Comment by mathematicsofpaul** [comment]
So just to get this straight, does your times series package allow me to input `[x_1, x_2. x_3, x_4]` and output `[x_5, x_6, x_7]` for the purpose of time series prediction? 

p.s I might be asking quite a few questions coming up as me a few colleagues are interested in trying this out :)

**Comment by maxjcohen** [comment]
Hi, neither the original transformer, nor my slightly modified version encourages that kind of usage. We focus on coherent many to many problems: for each input, the model predicts one (and only one) output. In other words, for your input `[x_1, x_2, x_3, x_4]`, the transformer would output `[y_1, y_2, y_3, y_4]`.
Additionally, because the original transformer model has a quadratic complexity in the time dimension, prediction problems are not usually suited.

If you have other questions, I'll be happy to discuss them ! Don't forget to open a new issue if it's not related to this one.

**Comment by mathematicsofpaul** [comment]
When you mention 

> the model predicts one (and only one) output 

but wrote `[y_1, y_2, y_3, y_4]`. Did you mean `[y_1]` only? 

Secondly, could say more about what you mean by the use of the word "coherent"? 

Thanks again!

**Comment by maxjcohen** [comment]
No, I mean exactly `[y_1, y_2, y_3, y_4]`, this is why it's a "many to many" model, and not a "many to one". When quoting my response, you omitted a key part:

> **for each input**, the model predicts one (and only one) output.

Coherent means (in this context) that for each input, we have exactly one output. This distinction aims a separating models like the transformer from other "many to many" models, that could for example input `[x_1, x_2, x_3, x_4]` and output `[y_5, y_6, y_7, y_8, y_9, ...]`. The key concept of "coherent many to many" is that each output corresponds to a named input.

**Comment by mathematicsofpaul** [comment]
I see, so it outputs outputs in unison as opposed to one after the other? Okay, so just to be sure, does your **decoder** attend to values like `[y_1, y_2, y_3]` when calculating `[y_4]`?

**Comment by maxjcohen** [comment]
Yes that's a good way of putting it.

At time step `k`, the decoder predicting `[y_k]` attends to the _latent vectors_ corresponding to outputs `[y_s, s<=k]`. For more detail, see the [original paper](https://arxiv.org/abs/1706.03762) as well as the [paper from this repo](https://arxiv.org/abs/2006.12390).
